src/ucp/rndv/rndv.c

```diff
@@ -949,10 +954,86 @@ ucp_rndv_send_frag_update_get_rkey(ucp_worker_h worker, ucp_request_t *freq,
     memset(rkey_index, 0, UCP_MAX_LANES * sizeof(uint8_t));
 }
 
+ucs_mpool_ops_t ucp_frag_mpool_ops = {
+    .chunk_alloc   = ucp_frag_mpool_malloc,
+    .chunk_release = ucp_frag_mpool_free,
+    .obj_init      = ucp_frag_mpool_obj_init,
+    .obj_cleanup   = ucs_empty_function
+};
+
+static ucs_mpool_t *
+ucp_rndv_get_mpool(ucp_worker_h worker, const ucp_worker_mpool_key_t *key)
+{
+    int num_frags    = 128;
+    size_t frag_size = worker->context->config.ext.rndv_frag_size[key->mem_type];
+    ucs_mpool_t *mpool;
+    khiter_t khiter;
+    int khret;
+    ucp_rndv_mpool_priv_t *mpriv;
+    ucs_status_t status;
+
+    ucs_recursive_spin_lock(&worker->mpool_hash_lock);
```

Comment:
would not this be called from progress only? Then this lock looks excessive

Reply:
@brminich Can multiple threads which share worker not cause a problem if they enter this function when worker is created with UCS_THREAD_MODE_MULTI?

Reply:
No, because in MT mode workers are supposed to be protected by a global lock taken in every UCP API function.